### ISSUE TITLE
[REVIEW]: FitBenchmarking: an open source Python package comparing data fitting software

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3127]
**Submitting author:** @tyronerees (<a href="http://orcid.org/0000-0003-0476-2259">Tyrone Rees</a>)
**Repository:** <a href="https://github.com/fitbenchmarking/fitbenchmarking" target ="_blank">https://github.com/fitbenchmarking/fitbenchmarking</a>
**Version:** v0.1.4
**Editor:** @dhhagan
**Reviewer:** @johnsamuelwrites, @djmitche
**Archive:** <a href="https://doi.org/10.5281/zenodo.4892752" target="_blank">10.5281/zenodo.4892752</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/ab5941478736701afc74fa4ce8ec6e12/status.svg)](https://joss.theoj.org/papers/ab5941478736701afc74fa4ce8ec6e12)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/ab5941478736701afc74fa4ce8ec6e12"><img src="https://joss.theoj.org/papers/ab5941478736701afc74fa4ce8ec6e12/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/ab5941478736701afc74fa4ce8ec6e12/status.svg)](https://joss.theoj.org/papers/ab5941478736701afc74fa4ce8ec6e12)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/fitbenchmarking/fitbenchmarking" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@johnsamuelwrites & @djmitche, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @dhhagan know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fitbenchmarking/fitbenchmarking">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@tyronerees) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fitbenchmarking/fitbenchmarking">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@tyronerees) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/s10589-014-9687-3 is OK
- 10.5281/zenodo.3930098 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=1.26 s (146.5 files/s, 20268.3 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          90           2326           4330           7187
CSS                              4             25             42           4831
SVG                              1              0              0           2671
reStructuredText                43            823            428           1471
HTML                             8              2              0            383
Markdown                        10             56              0            158
Bourne Shell                    10             48             39            152
YAML                             2             12              0            147
Dockerfile                       2             41             26             94
JSON                             7              0              0             87
TeX                              1             10              0             74
INI                              3             22              0             50
DOS Batch                        1              8              1             27
make                             1              4              6             10
XML                              1              0              0              6
JavaScript                       1              0              0              1
-------------------------------------------------------------------------------
SUM:                           185           3377           4872          17349
-------------------------------------------------------------------------------


Statistical information for the repository '3a997573fc70551e14f63439' was
gathered on 2021/03/23.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Anders Markvardsen              56          1732           8905            5.38
Anders-Markvardsen              34           496           1120            0.82
Andrew Lister                  255          9933          48921           29.75
Andrew McCluskey                10           297            468            0.39
AndrewLister-STFC                2             2              2            0.00
Anthony Lim                     19           723            339            0.54
Anton Piccardo-Selg              2             6              3            0.00
Atom Anuchitanukul              15             2             28            0.02
AtomAnu                        121         50466           2162           26.60
Federico Montesino P            24          1895            616            1.27
Karl Palmen                      1           434             14            0.23
Michael Wathen                 469         17191          13207           15.36
Patrick                         81          2755           1845            2.32
Patrick Odagiu                 200         15356          10962           13.30
Pete Peterson                    1             2              2            0.00
Roman Tolchenov                  3             3              3            0.00
Ross Whitfield                   1            26             12            0.02
Simon Fernandes                 32           222            136            0.18
TOFarmer                         7           143             99            0.12
Tom Perkins                      1             1              1            0.00
Tyrone Rees                    117          1504            558            1.04
bb511                           68          3284           1743            2.54
jess-farmer                      4           193             48            0.12

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Anders Markvardsen          142            8.2         12.6               17.61
Andrew Lister              4627           46.6         13.3               15.00
Andrew McCluskey              9            3.0         15.9                0.00
Anthony Lim                   2            0.3         43.6                0.00
Anton Piccardo-Selg           1           16.7         49.2                0.00
AtomAnu                      55            0.1         20.1                7.27
Michael Wathen             7732           45.0          9.0               14.12
Patrick                      83            3.0         31.3               22.89
Ross Whitfield                1            3.8         49.1                0.00
TOFarmer                     38           26.6         15.9                7.89
Tyrone Rees                 881           58.6          9.9               13.96
bb511                        84            2.6         26.8               11.90
jess-farmer                 189           97.9          0.1               13.23
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03127/joss.03127/10.21105.joss.03127.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03127/joss.03127/10.21105.joss.03127.pdf) :page_facing_up: :point_left:

**Comment by johnsamuelwrites** [comment]
@tyronerees Is it possible to add a section 'State of the field' in the article?

"State of the field: How FitBenchmarking software compares to other commonly-used packages?"

**Comment by tyronerees** [comment]
@johnsamuelwrites -- of course: I'll add something on that to the article

**Comment by djmitche** [comment]
"State of the Field" is the only checkbox left for me.  This looks like a well-managed OSS project, and spot-checks of the code all look well-organized and approachable.

**Comment by johnsamuelwrites** [comment]
I covered all the items in the review checklist. Overall, fitbenchmarking is a well-documented software with commits from multiple contributors. The codebase contains necessary test cases and continuous integration workflows. The different modules are adequately structured and well-commented/well-documented. Thanks to the examples, new users can quickly start working and testing the software. 

Further scope for improvement:
The documentation for the complete installation of the software with external dependencies for testing of all the functionalities needs some improvement. Currently, the users/testers may need to refer ".github/workflows/main.yml".

Also, please refer to my previous comment on "State of the field".

**Comment by whedon** [comment]
:wave: @johnsamuelwrites, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @djmitche, please update us on how your review is going (this is an automated reminder).

**Comment by kyleniemeyer** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03127/joss.03127/10.21105.joss.03127.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03127/joss.03127/10.21105.joss.03127.pdf) :page_facing_up: :point_left:

**Comment by kyleniemeyer** [comment]
Hi @tyronerees, it looks like we may be waiting on you to make some edits to the article?

**Comment by tyronerees** [comment]
Thanks for the prompt, @kyleniemeyer -- we've just merged in an updated article with a state of the field section added.  

@whedon generate pdf

**Comment by tyronerees** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03127/joss.03127/10.21105.joss.03127.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03127/joss.03127/10.21105.joss.03127.pdf) :page_facing_up: :point_left:

**Comment by kyleniemeyer** [comment]
Great, thanks @tyronerees! One final thing I noticed: for the Arnold et al. reference, you give a URL (that I'm not sure I think is appropriate, if that is a journal citation), but is there a DOI for that paper?

After making that change, can you archive the software repository (e.g., on Zenodo), and provide the DOI here? That is the last thing we need before accepting.

**Comment by johnsamuelwrites** [comment]
Thanks @tyronerees for adding "State of the field". I have now updated the checklist. Apart from the comment from @kyleniemeyer above, everything else looks fine.

**Comment by dhhagan** [comment]
Hey @tyronerees - as mentioned above, once you fix the reference mentioned by @kyleniemeyer above, could you please complete the following:

- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by kthyng** [comment]
Hi @tyronerees just a gentle ping that we are waiting on you here.

**Comment by djmitche** [comment]
@tyronerees how is this going?

**Comment by tyronerees** [comment]
Hi @djmitche -- this should be ready now (apologies for the delay -- we had some updates we wanted to get into the next release, and then I took my eye off the ball)

The doi is https://doi.org/10.5281/zenodo.4892752

Please let me know if you need me to do anything further.

**Comment by djmitche** [comment]
Looks like the release was 0.1.4, on zenodo.  Looks good as far as I know -- but I'm just the reviewer, not the editor :)

**Comment by kthyng** [comment]
@tyronerees Can you verify the version associated with this publication?

**Comment by kthyng** [comment]
Zenodo archive looks good ✅

**Comment by kthyng** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03127/joss.03127/10.21105.joss.03127.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03127/joss.03127/10.21105.joss.03127.pdf) :page_facing_up: :point_left:

**Comment by kthyng** [comment]
Paper looks good!

**Comment by tyronerees** [comment]
@kthyng -- the version that was first reviewed was 0.1.0.  The version with the changes to the paper requested and the zenodo release (with correct metadata) is 0.1.4.  No new features added between these releases -- just a couple of bug fixes and documentation tweaks.  

Please let me know if you need anything else from us.

**Comment by dhhagan** [comment]
@whedon set https://doi.org/10.5281/zenodo.4892752 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.4892752" target="_blank">10.5281/zenodo.4892752</a> is the archive.

**Comment by dhhagan** [comment]
@whedon set v0.1.4 as version

**Comment by whedon** [comment]
OK. v0.1.4 is the version.

**Comment by dhhagan** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.nima.2014.07.029 is OK
- 10.1007/s10589-014-9687-3 is OK
- 10.5281/zenodo.3930098 is OK
- 10.1007/s10898-013-0131-5 is OK

MISSING DOIs

- 10.1088/2632-2153/abedc8 may be a valid DOI for title: Olympus: a benchmarking framework for noisy optimization and experiment planning

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2370

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2370, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kthyng** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1402274123741552641) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2371
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03127](https://doi.org/10.21105/joss.03127)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by kthyng** [comment]
Congrats on your new publication @tyronerees! Thanks to editor @dhhagan and reviewers @johnsamuelwrites and @djmitche for your time, hard work, and expertise!!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03127/status.svg)](https://doi.org/10.21105/joss.03127)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03127">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03127/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03127/status.svg
   :target: https://doi.org/10.21105/joss.03127
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03127/status.svg)](https://doi.org/10.21105/joss.03127)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by tyronerees** [comment]
That's great news -- very many thanks all!
